src/room-observer.ts

```diff
@@ -43,40 +65,67 @@ export default class RoomObserver {
     debug('announcing to portal: %s', this.roomKey);
     pull(
       this.rpc.tunnel.endpoints(),
-      (this.endpointsDrain = pull.drain((endpoints: Array<FeedId>) => {
-        const room = this.roomKey;
-        debug('got endpoints from %s: %s', room, JSON.stringify(endpoints));
+      (this.endpointsDrain = pull.drain(
+        this.endpointsUpdated,
+        this.endpointsEnded,
+      )),
+    );
+  }
 
-        // Update onlineCount metadata for this room
-        const onlineCount = endpoints.length;
-        this.ssb.conn.hub().update(this.address, {onlineCount});
+  private endpointsUpdated = (endpoints: Array<FeedId>) => {
+    const room = this.roomKey;
+    debug('got endpoints from %s: %s', room, JSON.stringify(endpoints));
 
-        // Detect removed endpoints, unstage them
-        for (const entry of this.ssb.conn.staging().entries()) {
-          const [addr, data] = entry;
-          if (data.room === room && data.key && !endpoints.includes(data.key)) {
-            debug('will conn.unstage("%s")', addr);
-            this.ssb.conn.unstage(addr);
-          }
-        }
+    // Update onlineCount metadata for this room
+    const onlineCount = endpoints.length;
+    this.ssb.conn.hub().update(this.address, {onlineCount});
+
+    // Detect removed endpoints, unstage them
+    for (const entry of this.ssb.conn.staging().entries()) {
+      const [addr, data] = entry;
+      if (data.room === room && data.key && !endpoints.includes(data.key)) {
+        debug('will conn.unstage("%s")', addr);
+        this.ssb.conn.unstage(addr);
+      }
+    }
 
-        // Stage all the new endpoints
-        for (const key of endpoints) {
-          if (key === room) continue;
-          if (key === this.ssb.id) continue;
-          if (this.isAlreadyConnected(key)) continue;
-          const address = this.getAddress(key);
-          debug('will conn.stage("%s")', address);
-          this.ssb.conn.stage(address, {
-            type: 'room-endpoint',
-            key,
-            room,
-            roomName,
-          });
+    // Stage all the new endpoints
+    for (const key of endpoints) {
+      if (key === room) continue;
+      if (key === this.ssb.id) continue;
+      if (this.isAlreadyConnected(key)) continue;
+      const address = this.getAddress(key);
+      debug('will conn.stage("%s")', address);
+      this.ssb.conn.stage(address, {
+        type: 'room-endpoint',
+        key,
+        room,
+        roomName: this.roomMetadata?.name,
+      });
+    }
+  };
+
+  private endpointsEnded = (err?: Error | true) => {
```

Comment:
@cryptix I mostly copied this from [ssb-replicate](https://github.com/ssbc/ssb-replicate/blob/7567b077ade9350e287ae8d3a08a84ae84d8498b/legacy.js#L272)

Reply:
I'm fine that we add this here, and then later into some reusable module :)